File: apps/ios/Riot/Apps/AppResourceResolver.swift
import Foundation

/// Holds one app's decoded bundle and serves its resources by exact path
/// match. iOS reaches resources over a custom `riot-app://<app_id_hex>/<path>`
/// scheme, so — unlike Android's synthetic-origin resolver — there is no
/// origin host to derive here; the app id is kept only so the scheme handler
/// can confirm a request's host belongs to this app.
public struct AppResourceResolver: Sendable {
    /// Lowercased hex app id. Browsers lowercase URL hosts, so comparisons in
    /// the scheme handler are against this normalized form.
    public let appIDHex: String
    public let entryPoint: String
    private let resources: [AppResource]

    public init(appIDHex: String, bundle: DecodedAppBundle) {
        self.appIDHex = appIDHex.lowercased()
        self.entryPoint = bundle.entryPoint
        self.resources = bundle.resources
    }

    /// Exact-match lookup is the traversal defense — no path interpretation
    /// happens at all; "../x" simply matches no resource.
    public func resolve(path: String) -> AppResource? {
        if path.isEmpty { return nil }
        return resources.first { $0.path == path }
    }
}

File: apps/android/app/src/main/kotlin/org/riot/evidence/apps/AppResourceResolver.kt
package org.riot.evidence.apps

/**
 * Derives a per-app synthetic https origin and serves the bundle's
 * resources by exact path match. The hex app id is split into two 32-char
 * DNS labels (labels cap at 63 chars), giving every app a distinct origin
 * on `riot-app.invalid`.
 */
class AppResourceResolver(appIdHex: String, private val bundle: DecodedAppBundle) {
    val originHost: String =
        "${appIdHex.substring(0, 32)}.${appIdHex.substring(32)}.riot-app.invalid"
    val entryUrl: String = "https://$originHost/${bundle.entryPoint}"

    /**
     * Exact-match lookup is the traversal defense — no path interpretation
     * happens at all; "../x" simply matches no resource.
     */
    fun resolve(host: String?, rawPath: String?): AppResource? {
        if (host != originHost || rawPath == null || !rawPath.startsWith("/")) return null
        val path = rawPath.substring(1)
        if (path.isEmpty()) return null
        return bundle.resources.firstOrNull { it.path == path }
    }
}
